File: backend/app/rules.py
from typing import List
from .config import settings
from .models import InvoiceLine, MatchStatus, RoleType


def expected_rate_for_role(role: RoleType) -> float:
    if role == RoleType.MAIN_OPERATOR:
        return settings.RATE_MAIN_OPERATOR
    if role == RoleType.SECOND_OPERATOR:
        return settings.RATE_SECOND_OPERATOR
    if role == RoleType.YARD:
        return settings.RATE_YARD
    if role == RoleType.TRAVEL_DRIVER:
        return settings.RATE_TRAVEL_DRIVER
    if role == RoleType.TRAVEL_PASSENGER:
        return settings.RATE_TRAVEL_PASSENGER
    return 0.0


def check_rates(line: InvoiceLine, issues: List[str]):
    expected_rate = expected_rate_for_role(line.role)
    if abs(line.rate_per_hour - expected_rate) > 0.01:
        issues.append(f"Rate mismatch: expected {expected_rate}, got {line.rate_per_hour}")


def check_yard_hours(line: InvoiceLine, issues: List[str]):
    if line.hours_yard > 0 and line.hours_yard > settings.YARD_DAY_MAX_HOURS + 0.1:
        issues.append(f"Yard hours exceed max {settings.YARD_DAY_MAX_HOURS}h")


def check_full_shift(line: InvoiceLine, issues: List[str]):
    if line.hours_on_site > 0 and abs(line.hours_on_site - settings.FULL_SHIFT_HOURS) > 0.5:
        issues.append(
            f"On-site hours {line.hours_on_site} differ from standard {settings.FULL_SHIFT_HOURS}h (±0.5h)"
        )


def check_maths(line: InvoiceLine, issues: List[str]):
    calc = (line.hours_on_site + line.hours_travel + line.hours_yard) * line.rate_per_hour
    if abs(calc - line.line_total) > 0.5:
        issues.append(f"Maths error: expected {calc:.2f}, got {line.line_total:.2f}")


def compute_status(issues: List[str], has_jobsheet: bool, has_yard_record: bool) -> MatchStatus:
    if "Missing job sheet" in issues or "No yard sign-in" in issues:
        return MatchStatus.NEEDS_REVIEW
    if issues:
        return MatchStatus.PARTIAL
    if has_jobsheet or has_yard_record:
        return MatchStatus.MATCHED
    return MatchStatus.NEEDS_REVIEW


def apply_rules(line: InvoiceLine, has_jobsheet: bool, has_yard_record: bool) -> None:
    issues: List[str] = []

    if not has_jobsheet and line.hours_on_site > 0:
        issues.append("Missing job sheet")
    if line.hours_yard > 0 and not has_yard_record:
        issues.append("No yard sign-in")

    check_rates(line, issues)
    check_yard_hours(line, issues)
    check_full_shift(line, issues)
    check_maths(line, issues)

    line.match_status = compute_status(issues, has_jobsheet, has_yard_record)
    line.match_notes = "; ".join(issues)
    line.match_score = max(0.0, 1.0 - 0.1 * len(issues))

File: src/rules.php
<?php
declare(strict_types=1);

function expected_rate_for_role(string $role, ?bool $hasHgv): float
{
    if ($role === 'travel_driver') {
        if ($hasHgv === true)  return RATES['travel_driver_hgv'];
        if ($hasHgv === false) return RATES['travel_driver_non_hgv'];
        return RATES['travel_driver'];
    }
    return RATES[$role] ?? 0.0;
}

/**
 * Apply rule checks to one invoice line.
 *
 * $context may include:
 *   - expected_round_trip_hours: float|null  (2 × ORS one-way)
 *   - travel_debug: string                   (free-form notes for UI)
 *   - travel_tolerance:           float
 */
function apply_rules(
    array &$line,
    bool $hasJobsheet,
    bool $hasYardRecord,
    ?bool $hasHgv,
    array $context = []
): void {
    $issues = [];

    if (!$hasJobsheet && $line['hours_on_site'] > 0) {
        $issues[] = 'Missing job sheet';
    }
    if ($line['hours_yard'] > 0 && !$hasYardRecord) {
        $issues[] = 'No yard sign-in';
    }

    // Travel-time check: compare claimed hours_travel (round-trip) against
    // 2 × ORS one-way estimate ± tolerance.
    $expected = $context['expected_round_trip_hours'] ?? null;
    $tol      = $context['travel_tolerance'] ?? TRAVEL_TOLERANCE_HOURS;
    if ($expected !== null && $line['hours_travel'] > 0) {
        $delta = $line['hours_travel'] - $expected;
        if (abs($delta) > $tol) {
            $issues[] = sprintf(
                'Travel mismatch: claimed %.2fh, expected %.2fh round-trip (Δ %+.2fh, tol %.1fh)',
                $line['hours_travel'], $expected, $delta, $tol
            );
        }
    } elseif ($expected === null && $line['hours_travel'] > 0 && $hasJobsheet) {
        $issues[] = 'Travel: no postcode on matched job sheet';
    }

    $expected = expected_rate_for_role($line['role'], $hasHgv);
    if ($expected > 0 && abs($line['rate_per_hour'] - $expected) > 0.01) {
        $issues[] = sprintf('Rate mismatch: expected %.2f, got %.2f', $expected, $line['rate_per_hour']);
    }

    if ($line['hours_yard'] > YARD_DAY_MAX_HOURS + 0.1) {
        $issues[] = sprintf('Yard hours exceed max %.1fh', YARD_DAY_MAX_HOURS);
    }

    if ($line['hours_on_site'] > 0 && abs($line['hours_on_site'] - FULL_SHIFT_HOURS) > 0.5) {
        $issues[] = sprintf(
            'On-site hours %.2f differ from standard %.1fh (±0.5h)',
            $line['hours_on_site'],
            FULL_SHIFT_HOURS
        );
    }

    $calc = ($line['hours_on_site'] + $line['hours_travel'] + $line['hours_yard']) * $line['rate_per_hour'];
    if (abs($calc - $line['line_total']) > 0.5) {
        $issues[] = sprintf('Maths error: expected %.2f, got %.2f', $calc, $line['line_total']);
    }

    $line['match_notes'] = implode('; ', $issues);
    $line['match_score'] = max(0.0, 1.0 - 0.1 * count($issues));

    if (in_array('Missing job sheet', $issues, true) || in_array('No yard sign-in', $issues, true)) {
        $line['match_status'] = 'NEEDS_REVIEW';
    } elseif ($issues) {
        $line['match_status'] = 'PARTIAL';
    } elseif ($hasJobsheet || $hasYardRecord) {
        $line['match_status'] = 'MATCHED';
    } else {
        $line['match_status'] = 'NEEDS_REVIEW';
    }
}
